zio-http/src/main/scala/zhttp/http/Middleware.scala

```diff
@@ -1,450 +1,301 @@
-    self combine other
-
-  final def apply[R1 <: R, E1 >: E](app: HttpApp[R1, E1]): HttpApp[R1, E1] = self.execute(app, Middleware.Flag())
-
-  final def as[R1 <: R, E1 >: E](app: HttpApp[R1, E1]): Middleware[R1, E1] =
-    Middleware.Constant(app)
-
-  final def combine[R1 <: R, E1 >: E](other: Middleware[R1, E1]): Middleware[R1, E1] =
-    Middleware.Combine(self, other)
-
-  final def delay(duration: Duration): Middleware[R with Clock, E] = {
-    self.modifyZIO((_, _, _) => UIO(self).delay(duration))
-  }
-
-  final def execute[R1 <: R, E1 >: E](app: HttpApp[R1, E1], flags: Flag): HttpApp[R1, E1] =
-    Middleware.execute(self, app, flags)
-
-  final def modify[R1 <: R, E1 >: E](f: RequestP[Middleware[R1, E1]]): Middleware[R1, E1] =
-    Middleware.fromMiddlewareFunction((m, u, h) => f(m, u, h))
-
-  final def modifyZIO[R1 <: R, E1 >: E](
-    f: RequestP[ZIO[R1, Option[E1], Middleware[R1, E1]]],
-  ): Middleware[R1, E1] =
-    Middleware.fromMiddlewareFunctionZIO((m, u, h) => f(m, u, h))
-
-  final def orElse[R1 <: R, E1](other: Middleware[R1, E1]): Middleware[R1, E1] =
-    Middleware.OrElse(self, other)
-
-  final def race[R1 <: R, E1 >: E](other: Middleware[R1, E1]): Middleware[R1, E1] =
-    Middleware.Race(self, other)
-
-  final def setEmpty(flag: Boolean): Middleware[R, E] = Middleware.EmptyFlag(self, flag)
-
-  final def when(f: RequestP[Boolean]): Middleware[R, E] =
-    modify((m, u, h) => if (f(m, u, h)) self else Middleware.identity)
-
-  final def withEmpty: Middleware[R, E] = self.setEmpty(true)
-
-  final def withoutEmpty: Middleware[R, E] = self.setEmpty(false)
-}
-
-object Middleware {
-
-  type RequestP[+A] = (Method, URL, Headers) => A
+sealed trait Middleware[-R, +E, +AIn, -BIn, -AOut, +BOut] { self =>
 
   /**
-   * Sets cookie in response headers
+   * Applies self but if it fails, applies other.
    */
-  def addCookie(cookie: Cookie): Middleware[Any, Nothing] =
-    Middleware.addHeader(Headers.setCookie(cookie))
+  final def <>[R1 <: R, E1, AIn0 >: AIn, BIn0 <: BIn, AOut0 <: AOut, BOut0 >: BOut](
+    other: Middleware[R1, E1, AIn0, BIn0, AOut0, BOut0],
+  ): Middleware[R1, E1, AIn0, BIn0, AOut0, BOut0] = self orElse other
 
   /**
-   * Adds the provided header and value to the response
+   * Composes one middleware with another.
    */
-  def addHeader(name: String, value: String): Middleware[Any, Nothing] =
-    patch((_, _) => Patch.addHeader(name, value))
+  final def <<<[R1 <: R, E1 >: E, A0 <: AOut, B0 >: BOut, A1, B1](
+    other: Middleware[R1, E1, A0, B0, A1, B1],
+  ): Middleware[R1, E1, AIn, BIn, A1, B1] = self compose other
 
   /**
-   * Adds the provided header to the response
+   * Combines two middleware into one.
    */
-  def addHeader(header: Headers): Middleware[Any, Nothing] =
-    patch((_, _) => Patch.addHeader(header))
+  final def ++[R1 <: R, E1 >: E, A0 >: AIn <: AOut, B0 >: BOut <: BIn](
+    other: Middleware[R1, E1, A0, B0, A0, B0],
+  ): Middleware[R1, E1, A0, B0, A0, B0] =
```

Comment:
We should probably add some constraint types here. I don't think it infers properly.